python/pyspark/sql/types.py

```diff
@@ -2707,7 +2707,9 @@ def new_name(n: str) -> str:
             return "field %s in %s" % (n, name)
 
     def verify_nullability(obj: Any) -> bool:
-        if obj is None:
+        if obj is None or (isinstance(obj, decimal.Decimal) and obj.is_nan()):
+            # Spark's DecimalType doesn't support NaN,
+            # casting DoubleType NaN to DecimalType will return Null.
```

Comment:
It seems to describe only `nullable` case. For non-nullable columns, we raise exceptions, don't we?

Reply:
Existing behavior is that:
```
In [10]: data = [Row(v=float("NaN"))]

In [11]: schema = StructType([StructField("v", DoubleType(), False)])

In [12]: df = spark.createDataFrame(data=data, schema=schema)

In [13]: df.select('*', df.v.cast("decimal")).show()
+---+----+
|  v|   v|
+---+----+
|NaN|NULL|
+---+----+


In [14]: df.select('*', df.v.cast("decimal")).printSchema()
root
 |-- v: double (nullable = false)
 |-- v: decimal(10,0) (nullable = true)
```

The double->decimal conversion always changes the nullability to true